fix(sae/rpc): restore evicted blocks for state, receipt, and tracing RPCs (#5721)
ava-labs/avalanchego: vms/saevm/sae/rpc/README.md

Reviewer: Something that I think would be helpful is a comparison of what the different tracer backends do. Maybe a table. Like do all of them fake the state root and base fee? How many of them apply the next block's before block hook?

Author: Your other comment suggestion for a table I like more than this suggestion and I made it.

base fee and no post-execution state root. RPCs that need state (`eth_call`,
`eth_getBalance`, `debug_trace*`, …) wait for execution and are served faked
headers carrying the executed base fee and post-execution root, mimicking a
synchronous chain.

The tracing endpoints of the `debug` namespace are served by `tracerAPI`, which
routes each call to a `tracers.API` instance over one of three wrappers around
`backend`, the adapter from the SAE `Chain` to libevm's API packages. Any method
a wrapper doesn't override falls through to the next layer, so only the methods
below vary:

| Method | Behavior | Implementation |
| --- | --- | --- |
| `StateAtBlock` | The state as of the block's own execution, applying **no** before-block changes. | `backend` |
| | The parent's post-execution state with the **canonical child's** before-block changes applied, because block tracing asks for the state the child's transactions ran on. | `tracerBackend` |
| | Straight to `backend`, so **no** before-block changes: `debug_traceCall` wants the state as of the block itself, not the base state its child's transactions ran on. | `traceCallBackend` |

> | | As `tracerBackend`, except the **caller-supplied block** supplies the before-block changes, since it need not be canonical. | `suppliedHashBackend` |

Reviewer: ```suggestion
| `StateAtBlock` | The post-execution state of the provided block. | `backend` |
| | The provided block's post-execution state with the **canonical child's** before-block changes applied, allowing transaction tracing on the child to include these operations. | `tracerBackend` |
| | The post-execution state of the provided block. `debug_traceCall` is expected to be executed as if it was the last transaction. | `traceCallBackend` |
| | The provided block's post-execution state with the **caller-supplied block** before-block changes. | `suppliedHashBackend` |
```
This reads too much into context. I think it's clearer if you avoid the circular references.

Probably the lessons here could be provided to other rows in the table

# RPC

This package serves all of the SAE VM's JSON-RPC APIs. It adapts the VM into
the backends expected by libevm's `ethapi`, `tracers`, and `filters` packages
and registers every namespace (`eth`, `debug`, `txpool`, `net`, `web3`, …) on a
single `rpc.Server`; see `server.go` for the full endpoint list.

## Stateful RPCs

> SAE executes blocks asynchronously, so a stored header carries the worst-case

Reviewer: ```suggestion
Now that blocks are asynchronously, a stored header carries the worst-case
```
That is what SAE means (don't actually apply this diff)

Author: What do you mean don't apply this dif?

Author: respolved

base fee and no post-execution state root. RPCs that need state (`eth_call`,
`eth_getBalance`, `debug_trace*`, …) wait for execution and are served faked
headers carrying the executed base fee and post-execution root, mimicking a
synchronous chain.

The tracing endpoints of the `debug` namespace are served by `tracerAPI`,
which routes each call to a `tracers.API` instance over one of three backends. All of them wrap
`backend`, and any method a wrapper doesn't override falls through to the next
layer:

> 

Reviewer: This section would be more clear if organized by method, and then behavior, and then the name of the backend implementation
